pkg/common/workapplier/workapplier.go

```diff
@@ -0,0 +1,178 @@
+			if err := workClient.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, work); err != nil {
+				return err
+			}
+			return workClient.Delete(ctx, work)
+		},
+		patchWork: func(ctx context.Context, namespace, name string, pt types.PatchType, data []byte) (*workapiv1.ManifestWork, error) {
+			work := &workapiv1.ManifestWork{}
+			if err := workClient.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, work); err != nil {
+				return nil, err
+			}
+			if err := workClient.Patch(ctx, work, client.RawPatch(pt, data)); err != nil {
+				return nil, err
+			}
+			if err := workClient.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, work); err != nil {
+				return nil, err
+			}
+			return work, nil
+		},
+		createWork: func(ctx context.Context, work *workapiv1.ManifestWork) (*workapiv1.ManifestWork, error) {
+			if err := workClient.Create(ctx, work); err != nil {
+				return nil, err
+			}
+			createdWork := &workapiv1.ManifestWork{}
+			err := workClient.Get(ctx, types.NamespacedName{Namespace: work.Namespace, Name: work.Name}, createdWork)
+			return createdWork, err
+		},
+	}
+}
+
+func (w *WorkApplier) Apply(ctx context.Context, work *workapiv1.ManifestWork) (*workapiv1.ManifestWork, error) {
+	existingWork, err := w.getWork(ctx, work.Namespace, work.Name)
+	existingWork = existingWork.DeepCopy()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			existingWork, err = w.createWork(ctx, work)
+			if err == nil {
+				w.cache.updateCache(work, existingWork)
+				return existingWork, nil
+			}
+			return nil, err
+		}
+		return nil, err
+	}
+
+	if w.cache.safeToSkipApply(work, existingWork) {
+		return existingWork, nil
+	}
+
+	if manifestWorkSpecEqual(work.Spec, existingWork.Spec) {
+		return existingWork, nil
+	}
+
+	oldData, err := json.Marshal(&workapiv1.ManifestWork{
+		Spec: existingWork.Spec,
+	})
+	if err != nil {
+		return existingWork, err
+	}
+
+	newData, err := json.Marshal(&workapiv1.ManifestWork{
+		ObjectMeta: metav1.ObjectMeta{
+			UID:             existingWork.UID,
+			ResourceVersion: existingWork.ResourceVersion,
+		},
+		Spec: work.Spec,
+	})
+	if err != nil {
+		return existingWork, err
+	}
+
+	patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
+	if err != nil {
+		return existingWork, fmt.Errorf("failed to create patch for addon %s: %w", existingWork.Name, err)
+	}
+
+	klog.V(2).Infof("Patching work %s/%s with %s", existingWork.Namespace, existingWork.Name, string(patchBytes))
+	updated, err := w.patchWork(ctx, existingWork.Namespace, existingWork.Name, types.MergePatchType, patchBytes)
+	if err == nil {
+		w.cache.updateCache(work, existingWork)
+		return updated, nil
+	}
+	return nil, err
+}
+
+func (w *WorkApplier) Delete(ctx context.Context, namespace, name string) error {
+	err := w.deleteWork(ctx, namespace, name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	w.cache.removeCache(name, namespace)
```

Comment:
maybe we miss it, but shouldn't we have lock here or inside the cache?

Reply:
added a sync.RWMutex in the cache